presto-main/src/main/java/com/facebook/presto/operator/ExchangeClientFactory.java

```diff
@@ -66,7 +67,11 @@ public ExchangeClientFactory(
         this.concurrentRequestMultiplier = concurrentRequestMultiplier;
         this.minErrorDuration = requireNonNull(minErrorDuration, "minErrorDuration is null");
         this.httpClient = requireNonNull(httpClient, "httpClient is null");
-        this.maxResponseSize = requireNonNull(maxResponseSize, "maxResponseSize is null");
+
+        requireNonNull(maxResponseSize, "maxResponseSize is null");
+        long maxResponseSizeBytes = Math.min(httpClient.getMaxContentLength(), maxResponseSize.toBytes()) / 2;
```

Comment:
Can you create a PR for that? We'd need to review it and see if that has any performance implications. Until then we will use this fix as a mitigation.